datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -529,6 +535,89 @@ fn type_union_resolution_coercion(
     }
 }
 
+pub fn try_type_union_resolution(data_types: &[DataType]) -> Result<Vec<DataType>> {
```

Comment:
As this is a public function we perhaps add some comments here (maybe highlighting how it is different than `type_union_resolution`)